plugins/core/hu.bme.mit.gamma.expression.model/src/hu/bme/mit/gamma/expression/util/ExpressionTypeDeterminator2.java

```diff
@@ -0,0 +1,403 @@
+		}
+		if (expression instanceof OpaqueExpression) {
+			// ?
+		}
+		if (expression instanceof UnaryPlusExpression) {
+			return getArithmeticUnaryType((UnaryPlusExpression) expression);
+		}
+		if (expression instanceof UnaryMinusExpression) {
+			return getArithmeticUnaryType((UnaryMinusExpression) expression);
+		}
+		if (expression instanceof SubtractExpression) {
+			return getArithmeticBinaryType((SubtractExpression) expression);
+		}
+		if (expression instanceof DivideExpression) {
+			return getArithmeticBinaryType((DivideExpression) expression);
+		}
+		if (expression instanceof ModExpression) {
+			return getArithmeticBinaryIntegerType((ModExpression) expression);
+		}
+		if (expression instanceof DivExpression) {
+			return getArithmeticBinaryIntegerType((DivExpression) expression);
+		}
+		if (expression instanceof AddExpression) {
+			return getArithmeticMultiaryType((AddExpression) expression);
+		}
+		if (expression instanceof MultiplyExpression) {
+			return getArithmeticMultiaryType((MultiplyExpression) expression);
+		}
+		if (expression instanceof ArrayAccessExpression) {
+			int depth = 0;
+			List<FieldDeclaration> fields = new ArrayList<FieldDeclaration>();
+			ReferenceExpression ref = (ArrayAccessExpression) expression;
+			TypeDefinition type;
+			while (true) {
+				if (ref instanceof ArrayAccessExpression) {
+					depth++;
+					ref = (ReferenceExpression)((ArrayAccessExpression)ref).getOperand();
+				} else if (ref instanceof RecordAccessExpression) {
+					RecordAccessExpression recordAccessExpression = (RecordAccessExpression) ref;
+					fields.add(0, recordAccessExpression.getFieldReference().getFieldDeclaration());
+					ref = (ReferenceExpression) recordAccessExpression.getOperand();
+				} else if (ref instanceof DirectReferenceExpression) {
+					type = ExpressionModelDerivedFeatures.getTypeDefinition(((DirectReferenceExpression)ref).getDeclaration().getType());
+					break;
+				} else {
+					throw new IllegalArgumentException("Array access expression contains forbidden elements: " + ref.getClass());
+				}
+			}
+			while (depth >= 0 || !fields.isEmpty()) {
+				if (type instanceof ArrayTypeDefinition) {
+					type = ExpressionModelDerivedFeatures.getTypeDefinition(((ArrayTypeDefinition)type).getElementType());
+					depth--;
+				} else if (type instanceof RecordTypeDefinition) {
+					type = ExpressionModelDerivedFeatures.getTypeDefinition(((RecordTypeDefinition)type).getFieldDeclarations().stream().filter(e -> e.equals(fields.remove(0))).findFirst().get().getType());
+				} else {
+					throw new IllegalArgumentException("Type contains forbidden elements: " + type.getClass());
+				}
+			}
+			return ExpressionModelDerivedFeatures.getTypeDefinition(type);
+		}
+		if (expression instanceof FunctionAccessExpression) {
+			Type declarationType = expressionUtil.getDeclaration(expression).getType();
+			return ExpressionModelDerivedFeatures.getTypeDefinition(declarationType);
```

Comment:
Let us talk about the use of type `ExpressionModelDerivedFeatures.getTypeDefinition(...)`, we should keep the retrieval of Types and TypeDefinitions separate.